index.js

```diff
@@ -316,6 +315,94 @@ const legacyScriptingRunner = (Zap, zcli, input) => {
     return _.template(templateString, options)(values);
   };
 
+  const ensureIsType = (result, type) => {
+    if (type) {
+      if (type.startsWith('array-')) {
+        if (Array.isArray(result)) {
+          return result;
+        } else if (result && typeof result === 'object') {
+          if (type === 'array-wrap') {
+            return [result];
+          } else {
+            // Find the first array in the response
+            for (const k in result) {
+              const value = result[k];
+              if (Array.isArray(value)) {
+                return value;
+              }
+            }
+          }
+        }
+        throw new Error('JSON results array could not be located.');
+      } else if (type.startsWith('object-')) {
+        if (_.isPlainObject(result)) {
+          return result;
+        } else if (
+          Array.isArray(result) &&
+          result.length > 0 &&
+          _.isPlainObject(result[0])
+        ) {
+          // Used by auth test and auth label
+          return result[0];
+        }
+        throw new Error('JSON results object could not be located.');
+      }
+    }
+    return result;
+  };
+
+  const isObject = (object) => {
+    return !Array.isArray(object) && typeof object === 'object';
+  };
+
+  const handleLegacyParams = (data) => {
+    // see handle_legacy_params in the python backend
+    if (!isObject(data)) {
+      return data;
+    }
+    var params = {};
+    for (var i in data) {
+      if (isObject(data[i])) {
+        var param = [];
+        for (var j in data[i]) {
+          param.push(j + ',' + data[i][j]);
+        }
+        params[i] = param.join('|');
+      } else if (Array.isArray(data[i])) {
+        params[i] = data[i].join();
+      }
+    }
+    return params;
+  };
+
+  const textifyList = (data) => {
+    // see textify_list in the python backend
+    if (!Array.isArray(data)){
+      return data;
+    }
+    var out = '';
```

Comment:
you could make this block less error prone if you make `out` an array and then `out.join(sep)` at the end. that way, you don't need to include `+ sep` in a bunch of places. Also (maybe) don't need that final regex. 

If you keep it as a string, it should be `let` and if it's an array, it can be `const`.

Reply:
I'm going to leave this as a string, but only because the string version lets us accurately recreate some subtle (most likely buggy) behaviour of the original python method. I tried to make it array based when I first wrote it but there were some odd things with the double newlines so I kept it as close to the original logic as possible. I've fixed the `var` -> `let` keyword though.